### PR TITLE
Using tag condition on get object

### PR DESCRIPTION
**Pull request by ftravaglia**
## Description
Implementing the use of "s3:ExistingObjectTag" policy conditions on GetObject

## Motivation and Context
Within a bucket, we want to give the possibility to give access or not on an object in reading for a given user. 

For example: If one of the files has the tag 'type:sensitive' and in the iam policies file the policy condition is set to Allow "s3:ExistingObjectTag/type": "sensitive" then the user will not be able to read the file nor download it (it remains visible in the bucket but only at the name level)

Translated with www.DeepL.com/Translator (free version)

## How to test this PR?
- deploy minio
- add policy 'testtagpolicy' 
`{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Action": [
                "s3:*"
            ],
            "Resource": [
                "arn:aws:s3:::*"
            ]
        },
        {
            "Effect": "Deny",
            "Action": [
                "s3:GetObject"
            ],
            "Resource": [
                "arn:aws:s3:::mybucket/*"
            ],
            "Condition": {
                "StringEquals": {
                    "s3:ExistingObjectTag/type": [
                        "sensitive"
                    ]
                }
            }
        }
    ]
}`
- as admin create user 'test'
- as admin link the user 'test' with the 'testtagpolicy'
- as admin create bucket 'mybucket'
- as admin go into the bucket and push a file
- as admin add tag to the file with key 'type' and value 'sensitive'
- as admin add another file without tag
- as test user login to minio
- as test go browse 'mybucket'
- as test try to download the file with tags
- as test try to download the file without tags

Result expected : 
The file with tags can't be downloaded (on firefox it download a file but with 0 bytes)
The file without tags can be downloaded

- as admin remove tag to the file with tags

Result expected : 
The file with tags removed can be downloaded or see into the minio console view

- as admin add tag to the file with key 'type' and value 'sensitive'
The file with tags can't be downloaded (on firefox it download a file but with 0 bytes)

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Documentation updated
- [ ] Unit tests added/updated

See [https://github.com/minio/minio/issues/10265](https://github.com/minio/minio/issues/10265)

### CODE REVIEW

File: cmd/object-handlers.go [hunk 1]
```diff
@@ -600,8 +600,13 @@ func (api objectAPIHandlers) headObjectHandler(ctx context.Context, objectAPI Ob
 		writeErrorResponseHeadersOnly(w, toAPIError(ctx, err))
 		return
 	}
-
-	if s3Error := checkRequestAuthType(ctx, r, policy.GetObjectAction, bucket, object); s3Error != ErrNone {
+	objAPI := api.ObjectAPI()
+	tags, err := objAPI.GetObjectTags(ctx, bucket, object, opts)
```
Comment (klauspost) [inline cmd/object-handlers.go:604]: Fetching tags, just to fetch object info following is effectively doubling the operations needed.

Fetch the information once.

File: cmd/object-handlers.go [hunk 2]
```diff
@@ -600,8 +600,13 @@ func (api objectAPIHandlers) headObjectHandler(ctx context.Context, objectAPI Ob
...
+	if err != nil {
+		writeErrorResponseHeadersOnly(w, toAPIError(ctx, err))
```
Comment (klauspost) [inline cmd/object-handlers.go:606]: This changes the order of errors and likely changes the errors returned when multiple problems exist.

In this case it seems that we "leak" whether the object exists, before we have check auth. This needs careful implementation.

See also `// As per "Permission" section in....` below.

**Review by klauspost — COMMENTED** [review]
Thanks for the contribution. Some initial observations on your draft.

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Comment by harshavardhana** [comment]
This PR is an incorrect way to implement this @ftravaglia

**Comment by ftravaglia** [comment]
> This PR is an incorrect way to implement this @ftravaglia

@harshavardhana Okay, so can you tell me the correct way to correct this, please? Thanks

**Comment by harshavardhana** [comment]
> > This PR is an incorrect way to implement this @ftravaglia
> 
> @harshavardhana Okay, so can you tell me the correct way to correct this, please? Thanks

First of all 

- Attempting to read tags is wrong at the handler layer this was on purpose not implemented, we cannot loose IOPs the way you have implemented.

- The clean way to implement this to implement this as a callback, i.e tagging based filtering has to be applied in the same manner as Get and Head based pre-conditions are applied for Range, Etag(match).. are implemented.

- No new package dependencies are entertained, such as another ftravaglia/pkg 

If you can make all these deeper changes you are free to attempt this change.

But honestly this is a more involved change than you have led yourself on, so my suggestion would be simply close this PR and wait for us to implement it cleanly.

**Comment by harshavardhana** [comment]
Closed this since this is not a correct change, please re-submit when the mentioned comments are addressed or wait for us to fix it.

**Pull request closed without merging**